src/knx/secure_application_layer.h

```diff
@@ -0,0 +1,277 @@
+ * methods of the BusAccessUnit class.
+ */
+class SecureApplicationLayer :  public ApplicationLayer
+{
+
+  public:
+    /**
+     * The constructor.
+     * @param assocTable The AssociationTable is used to translate between asap (i.e. group objects) and group addresses.
+     * @param bau methods are called here depending of the content of the APDU
+     */
+    SecureApplicationLayer(DeviceObject& deviceObj, SecurityInterfaceObject& secIfObj, AssociationTableObject& assocTable, BusAccessUnit& bau);
+
+    void setSecurityMode(bool enabled);
+    bool isSecurityModeEnabled();
+    void clearFailureLog();
+    void getFailureCounters(uint8_t* data);
+    uint8_t getFromFailureLogByIndex(uint8_t index, uint8_t* data, uint8_t maxDataLen);
+
+    // from transport layer
+    virtual void dataGroupIndication(HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu) override;
+    virtual void dataGroupConfirm(AckType ack, HopCountType hopType, Priority priority, uint16_t tsap,
+                                  APDU& apdu, bool status) override;
+    virtual void dataBroadcastIndication(HopCountType hopType, Priority priority, uint16_t source, APDU& apdu) override;
+    virtual void dataBroadcastConfirm(AckType ack, HopCountType hopType, Priority priority, APDU& apdu, bool status) override;
+    virtual void dataSystemBroadcastIndication(HopCountType hopType, Priority priority, uint16_t source, APDU& apdu) override;
+    virtual void dataSystemBroadcastConfirm(HopCountType hopType, Priority priority, APDU& apdu, bool status) override;
+    virtual void dataIndividualIndication(HopCountType hopType, Priority priority, uint16_t source, APDU& apdu) override;
+    virtual void dataIndividualConfirm(AckType ack, HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu, bool status) override;
+    virtual void dataConnectedIndication(Priority priority, uint16_t tsap, APDU& apdu) override;
+    virtual void dataConnectedConfirm(uint16_t tsap) override;
+
+    void loop();
+
+  protected:
+    // to transport layer
+    virtual void dataGroupRequest(AckType ack, HopCountType hopType, Priority priority, uint16_t tsap, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataBroadcastRequest(AckType ack, HopCountType hopType, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataSystemBroadcastRequest(AckType ack, HopCountType hopType, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataIndividualRequest(AckType ack, HopCountType hopType, Priority priority, uint16_t destination, APDU& apdu, const SecurityControl& secCtrl) override;
+    virtual void dataConnectedRequest(uint16_t tsap, Priority priority, APDU& apdu, const SecurityControl& secCtrl) override; // apdu must be valid until it was confirmed
+
+  private:
+
+    template <typename K, typename V, int SIZE>
```

Comment:
We should probably put this template in its own file.

Reply:
yep, will move it "SimpleMap.h".

Reply:
done.